### ISSUE TITLE
Scope of Retry-After

### ISSUE THREAD

**Issue by martinthomson** [httpwg/http-core#99]
Several recent discussions have raised the question of the scope of applicability of Retry-After.  It was clear from those discussions that some people interpreted the instruction as applying to the entire server.  Others take it to mean the request, others the resource.  It might also be possible to interpret the code as applying to all requests on a given connection, but that's not really consistent with the overall protocol design.

My understanding is that it can only apply to the request that is responding to, or at most the resource.  Requests are independent of each other and so - while a client might extend the scope by inference, or knowledge about the server - the only scope that the protocol can definitively say it applies to is either the request or the resource.

If it applies to the resource, it might be narrowed in scope to also apply to the method (POST is too expensive, but GET is fine, for example).

**Comment by mnot** [comment]
Since it's already deployed, we can't retroactively define a scope; best we can do is define a new protocol element for those who wish to convey an explicit scope.

**Comment by martinthomson** [comment]
Well, we might reasonably assume that the scope at least includes the request.  It's probably not talking about another request, but yeah, more work needed.

**Comment by mnot** [comment]
I can see pointing out what needs to be defined, and that the core spec doesn't provide that. Do you think we can do anything else in core?

**Comment by martinthomson** [comment]
The scope of "this request" is essentially meaningless.  Because when I make a subsequent request, that's a different request, even if the bits are all the same.  So I think that what we have is:

When a server provides a `Retry-After` header field in a response, it is an indication that making a similar request prior to the elapsed time is unlikely to result in a different answer. Clients are therefore requested to refrain from repeating the request within that time.

A `Retry-After` value can only apply to the specific request that it was sent in response to, but some conditions that lead to a server are necessarily broader in scope than a single request.  A server might include information in the response body to request that the client also refrain from making other requests to the same resource, or even to all resources on the same server.

Clients can apply heuristics to infer the scope that `Retry-After` applies to.  For instance, if a server indicates excessive load with a 503 (Service Unavailable) status code, a client might infer that the condition applies to more than one resource.

**Comment by ioggstream** [comment]
> Several recent discussions have raised the question of the scope of applicability of Retry-After

@martinthomson Can you please link some of them in the issue? 

> best we can do is define a new protocol element for those who wish to convey an explicit scope.

imho adding stuff will make things more confused. 

> A Retry-After value can only apply to the specific request that it was sent in response to

I agree. Gateways may amplify this behavior as they can proxy different services behind the same Origin,
so I can get a 503 from an URL but not from another one.

**Comment by spaceone** [comment]
> > Several recent discussions have raised the question of the scope of applicability of Retry-After
> 
> @martinthomson Can you please link some of them in the issue?

I can tell you oen of my use cases as one example for that I understand the scope is for the actual request:
I am using `Retry-After` in combination with a redirection containing a `Location` header to realize something like a progress bar.
The request flow looks like:

1. starts an asynchronous task, redirects to a resource containing the state of the task
```http
PUT /object/1?new_location=/object/2 HTTP/1.1
```
→
```http
HTTP/1.1 201 Accepted
Location: /progress/1
```

2. the state resource is a redirection loop until until the task is completed
```http
GET /progress/1 HTTP/1.1
```
→
```http
HTTP/1.1 303 See Other
Location: /progress/1
Retry-After: 1

<html>Moved 6 of 20 sub objects. Last object was: ...</html>
```
3. when the task has completed a redirection to the finished resource is done
```http
GET /progress/1 HTTP/1.1
```
→
```http
HTTP/1.1 303 See Other
Location: /object/2
```

If you have plans to introduce another header please consider that it would be nice if the delay could be given with milliseconds, e.g. `Retry: 0.6`.

**Comment by ioggstream** [comment]
### about a new header with millisecond precision

imho considering network and processing latency, subsecond precision brings more complexities than benefits, but please I'd really like to ear from you @mnot @martinthomson  

### about retry-after scopes

we could use link relations or the `Warning` header, defining a new `warn-code` with syntax+semantic.

**Comment by mnot** [comment]
Looking into this, I think `Retry-After` is pretty clear about its scope; it is specific to each use case. 

* For 503, the status code says that it's how long the client should wait before retrying the request
* For 3xx, it says it's about how long the client should wait before issuing the redirected request

At most, I could see adding a statement to the definition of 503 (NOT Retry-After) that clarifies that it's not necessarily about the state of the entire server -- but it could be, and clients can take that on board if they want to.

Thoughts?

**Comment by mnot** [comment]
@martinthomson my inclination is to close with no action, as per above. If you'd like clarifying text in 503, or something else, please say so.

**Comment by ioggstream** [comment]
Ok for clarifying text in 503.
We may suggest that `the scope MAY be conveyed in the payload body`.
This won't fix this issue though.

@martinthomson wdyt?

**Comment by martinthomson** [comment]
Clarifying 503 seems tolerable.

**Comment by mnot** [comment]
Based on the discussion in #317, I think this should be closed with no action. The scope of the 503 is already well-defined, and can be extended in the future.

**Comment by ioggstream** [comment]
Here's the proposal following on Roy's suggestion  in https://github.com/httpwg/http-core/pull/317#issuecomment-585868767 

I created a short extension draft https://ioggstream.github.io/draft-polli-retry-scope/draft-polli-retry-scope.html and closed #317 

Feedback welcome.
